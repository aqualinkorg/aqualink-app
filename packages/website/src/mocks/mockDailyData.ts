import { DailyData } from "../store/Reefs/types";

const now = new Date();
const minutesAgo = 5;
const dailyDataDate = new Date(
  now.getTime() - minutesAgo * 60000
).toISOString();

export const mockDailyData: DailyData = {
  id: 1,
  date: dailyDataDate,
  minBottomTemperature: 25,
  maxBottomTemperature: 27,
  avgBottomTemperature: 26,
  degreeHeatingDays: 29,
  surfaceTemperature: 36,
  satelliteTemperature: 20,
  minWaveHeight: 1,
  maxWaveHeight: 3,
  avgWaveHeight: 2,
  waveDirection: 136,
  wavePeriod: 15,
  minWindSpeed: 3,
  maxWindSpeed: 5,
  avgWindSpeed: 4,
  windDirection: 96,
  weeklyAlertLevel: 3,
};
