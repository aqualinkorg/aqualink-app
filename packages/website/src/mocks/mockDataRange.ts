import { TimeSeriesDataRange } from 'store/Sites/types';

export const mockDataRange: TimeSeriesDataRange = {
  bottomTemperature: [
    {
      type: 'hobo',
      depth: null,
      data: [
        {
          minDate: '2017-11-02T15:00:00.000Z',
          maxDate: '2020-02-02T21:00:00.000Z',
        },
      ],
    },
  ],
  cholorophyllConcentration: [
    {
      type: 'sonde',
      depth: null,
      data: [
        {
          minDate: '2017-11-02T15:00:00.000Z',
          maxDate: '2020-02-02T21:00:00.000Z',
        },
      ],
    },
  ],
  odoConcentration: [
    {
      type: 'sonde',
      depth: null,
      data: [
        {
          minDate: '2017-11-02T15:00:00.000Z',
          maxDate: '2020-02-02T21:00:00.000Z',
        },
      ],
    },
  ],
  ph: [
    {
      type: 'sonde',
      depth: null,
      data: [
        {
          minDate: '2017-11-02T15:00:00.000Z',
          maxDate: '2020-02-02T21:00:00.000Z',
        },
      ],
    },
  ],
  salinity: [
    {
      type: 'sonde',
      depth: null,
      data: [
        {
          minDate: '2017-11-02T15:00:00.000Z',
          maxDate: '2020-02-02T21:00:00.000Z',
        },
      ],
    },
    {
      type: 'hui',
      depth: null,
      data: [
        {
          minDate: '2022-06-11T00:00:00.000Z',
          maxDate: '2022-06-14T23:45:00.000Z',
        },
      ],
    },
  ],
  turbidity: [
    {
      type: 'sonde',
      depth: null,
      data: [
        {
          minDate: '2017-11-02T15:00:00.000Z',
          maxDate: '2020-02-02T21:00:00.000Z',
        },
      ],
    },
  ],
};
