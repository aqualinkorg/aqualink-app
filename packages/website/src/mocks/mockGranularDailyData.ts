import { DailyData } from 'store/Sites/types';

export const mockGranularDailyData: DailyData[] = [
  {
    id: 3,
    date: '2022-06-27T02:59:59.999Z',
    minBottomTemperature: 0,
    maxBottomTemperature: 0,
    avgBottomTemperature: 0,
    degreeHeatingDays: 0,
    topTemperature: 0,
    satelliteTemperature: 24.219252436179815,
    minWaveHeight: 0.17217956015502844,
    maxWaveHeight: 0.4220004534334179,
    avgWaveHeight: 0.28517965469411594,
    waveMeanDirection: 219,
    waveMeanPeriod: 2,
    minWindSpeed: 1.0233644143488965,
    maxWindSpeed: 5.558413659400355,
    avgWindSpeed: 2.9796803902255204,
    windDirection: 218,
    weeklyAlertLevel: 0,
  },
  {
    id: 2,
    date: '2022-06-26T02:59:59.999Z',
    minBottomTemperature: 0,
    maxBottomTemperature: 0,
    avgBottomTemperature: 0,
    degreeHeatingDays: 0,
    topTemperature: 0,
    satelliteTemperature: 24.229605847456323,
    minWaveHeight: 0.06706502244580449,
    maxWaveHeight: 0.40127747690701104,
    avgWaveHeight: 0.20286431769935753,
    waveMeanDirection: 186,
    waveMeanPeriod: 2,
    minWindSpeed: 1.0683246478923132,
    maxWindSpeed: 3.8642400151240097,
    avgWindSpeed: 2.4470019134494967,
    windDirection: 238,
    weeklyAlertLevel: 0,
  },
  {
    id: 1,
    date: '2022-06-25T02:59:59.999Z',
    minBottomTemperature: 0,
    maxBottomTemperature: 0,
    avgBottomTemperature: 0,
    degreeHeatingDays: 0,
    topTemperature: 0,
    satelliteTemperature: 24.00602354214886,
    minWaveHeight: 0.25399888336189147,
    maxWaveHeight: 0.7053866500812253,
    avgWaveHeight: 0.49466218581824045,
    waveMeanDirection: 184,
    waveMeanPeriod: 3,
    minWindSpeed: 2.3608638556613584,
    maxWindSpeed: 6.910529333003823,
    avgWindSpeed: 4.019357707456016,
    windDirection: 125,
    weeklyAlertLevel: 0,
  },
];
