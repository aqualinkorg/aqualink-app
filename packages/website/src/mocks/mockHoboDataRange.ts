import { TimeSeriesDataRange } from "../store/Reefs/types";

export const mockHoboDataRange: TimeSeriesDataRange = {
  hobo: {
    bottomTemperature: [
      {
        minDate: "2017-11-02T14:59:08.000Z",
        maxDate: "2020-02-02T20:59:08.000Z",
      },
    ],
    alert: [],
    dhw: [],
    satelliteTemperature: [],
    topTemperature: [],
    sstAnomaly: [],
    significantWaveHeight: [],
    wavePeakPeriod: [],
    waveMeanDirection: [],
    windSpeed: [],
    windDirection: [],
  },
  spotter: {
    bottomTemperature: [],
    alert: [],
    dhw: [],
    satelliteTemperature: [],
    topTemperature: [],
    sstAnomaly: [],
    significantWaveHeight: [],
    wavePeakPeriod: [],
    waveMeanDirection: [],
    windSpeed: [],
    windDirection: [],
  },
  sofarApi: {
    bottomTemperature: [],
    alert: [],
    dhw: [],
    satelliteTemperature: [],
    topTemperature: [],
    sstAnomaly: [],
    significantWaveHeight: [],
    wavePeakPeriod: [],
    waveMeanDirection: [],
    windSpeed: [],
    windDirection: [],
  },
};
