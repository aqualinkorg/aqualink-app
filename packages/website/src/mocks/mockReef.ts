import { CollectionMetrics, Point, Reef } from "../store/Reefs/types";
import { mockUser } from "./mockUser";

const now = new Date();
const minutesAgo = 5;
const liveDataDate = new Date(now.getTime() - minutesAgo * 60000).toISOString();

export const mockReef: Reef = {
  id: 1,
  name: "",
  polygon: {
    coordinates: [0, 0],
    type: "Point",
  },
  hasHobo: false,
  spotterId: null,
  maxMonthlyMean: 0,
  depth: 0,
  status: "in_review",
  videoStream: null,
  region: { name: "Hawaii" },
  admins: [mockUser],
  surveyPoints: [],
  stream: null,
  historicalMonthlyMean: [],
  liveData: {
    reef: { id: 1 },
    bottomTemperature: {
      value: 39,
      timestamp: liveDataDate,
    },
    satelliteTemperature: {
      value: 29,
      timestamp: liveDataDate,
    },
    degreeHeatingDays: {
      value: 34,
      timestamp: liveDataDate,
    },
    waveHeight: {
      value: 1,
      timestamp: liveDataDate,
    },
    waveDirection: {
      value: 90,
      timestamp: liveDataDate,
    },
    wavePeriod: {
      value: 3,
      timestamp: liveDataDate,
    },
    windSpeed: {
      value: 6,
      timestamp: liveDataDate,
    },
    windDirection: {
      value: 180,
      timestamp: liveDataDate,
    },
  },
  dailyData: [
    {
      id: 171,
      date: liveDataDate,
      minBottomTemperature: 37,
      maxBottomTemperature: 39,
      avgBottomTemperature: 38,
      degreeHeatingDays: 34,
      topTemperature: 29,
      satelliteTemperature: 23,
      minWaveHeight: 2,
      maxWaveHeight: 4,
      avgWaveHeight: 3,
      waveDirection: 205,
      wavePeriod: 28,
      minWindSpeed: 3,
      maxWindSpeed: 5,
      avgWindSpeed: 4,
      windDirection: 229,
    },
  ],
  latestDailyData: {
    id: 171,
    date: liveDataDate,
    minBottomTemperature: 37,
    maxBottomTemperature: 39,
    avgBottomTemperature: 38,
    degreeHeatingDays: 34,
    topTemperature: 29,
    satelliteTemperature: 23,
    minWaveHeight: 2,
    maxWaveHeight: 4,
    avgWaveHeight: 3,
    waveDirection: 205,
    wavePeriod: 28,
    minWindSpeed: 3,
    maxWindSpeed: 5,
    avgWindSpeed: 4,
    windDirection: 229,
  },
  collectionData: {
    topTemperature: 9.54,
    bottomTemperature: 10.719999999999999,
    satelliteTemperature: 15.8900003433228,
    degreeHeatingDays: 0,
    weeklyAlertLevel: 0,
    sstAnomaly: 1.2,
  },
};

export const generateMockReef = ({
  id,
  name,
  regionName,
  deployed,
  isCollection,
  historicMax,
  collectionData,
  polygon,
}: {
  id: number;
  name: string;
  regionName: string;
  deployed: boolean;
  isCollection: boolean;
  historicMax: number;
  collectionData: CollectionMetrics;
  polygon: Point;
}): Reef => ({
  ...mockReef,
  id,
  name,
  region: { ...mockReef.region, name: regionName },
  polygon,
  status: deployed ? "deployed" : mockReef.status,
  spotterId: deployed ? "SPOT-0930" : mockReef.spotterId,
  collectionData,
  latestDailyData: isCollection ? undefined : mockReef.latestDailyData,
  maxMonthlyMean: historicMax,
});
