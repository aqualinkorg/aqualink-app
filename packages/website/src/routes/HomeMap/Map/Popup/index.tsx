import React, { useEffect, useRef } from "react";
import {
  Button,
  Card,
  CardContent,
  CardHeader,
  createStyles,
  Grid,
  Theme,
  Tooltip,
  Typography,
  withStyles,
  WithStyles,
} from "@material-ui/core";
import { Link } from "react-router-dom";
import { Popup as LeafletPopup, useLeaflet } from "react-leaflet";
import { useSelector } from "react-redux";

import type { LatLngTuple } from "leaflet";
import type { Reef } from "../../../../store/Reefs/types";
import { getReefNameAndRegion } from "../../../../store/Reefs/helpers";
import { colors } from "../../../../layout/App/theme";
import { formatNumber } from "../../../../helpers/numberUtils";
import { dhwColorFinder } from "../../../../helpers/degreeHeatingWeeks";
import { reefOnMapSelector } from "../../../../store/Homepage/homepageSlice";

const Popup = ({ reef, classes, autoOpen }: PopupProps) => {
  const { map } = useLeaflet();
  const reefOnMap = useSelector(reefOnMapSelector);
  const popupRef = useRef<LeafletPopup>(null);
  const { name, region } = getReefNameAndRegion(reef);
  const isNameLong = name && name.length > 50;

  const { dhw, satelliteTemperature } = reef.collectionData || {};

  useEffect(() => {
    if (
      map &&
      popupRef?.current &&
      reefOnMap?.id === reef.id &&
      reefOnMap?.polygon.type === "Point" &&
      autoOpen
    ) {
      const { leafletElement: popup } = popupRef.current;
      const [lng, lat] = reefOnMap.polygon.coordinates;

      const point: LatLngTuple = [lat, lng];
      popup.setLatLng(point).openOn(map);
    }
  }, [autoOpen, map, reef.id, reefOnMap]);

  return (
    <LeafletPopup
      ref={reefOnMap?.id === reef.id ? popupRef : null}
      closeButton={false}
      className={classes.popup}
      autoPan={false}
    >
      <Card>
        <CardHeader
          className={classes.popupHeader}
          classes={{
            content: classes.popupHeaderContent,
            subheader: classes.subheader,
          }}
          title={
            <span title={isNameLong && name ? name : undefined}>
              {isNameLong ? `${name?.substring(0, 50)}...` : name}
            </span>
          }
          subheader={region}
        />
        <CardContent>
          <Grid container item xs={12}>
            <Grid item xs={6}>
              <Grid container item xs={12}>
                <Typography variant="caption" color="textSecondary">
                  SST
                </Typography>
              </Grid>
              <Grid container item xs={12}>
                <Typography
                  style={{ color: colors.lightBlue }}
                  variant="h5"
                  color="textSecondary"
                >
                  {`${formatNumber(satelliteTemperature, 1)}  Â°C`}
                </Typography>
              </Grid>
            </Grid>
            <Grid item xs={6}>
              <Grid container item xs={12}>
                <Typography variant="caption" color="textSecondary">
                  HEAT STRESS
                </Typography>
              </Grid>
              <Grid container alignItems="flex-end" item xs={12}>
                <Typography
                  style={{
                    color: `${dhwColorFinder(dhw)}`,
                  }}
                  variant="h5"
                  color="textSecondary"
                >
                  {formatNumber(dhw, 1)}
                  &nbsp;
                </Typography>
                <Tooltip title="Degree Heating Weeks - a measure of the amount of time above the 20 year historical maximum temperatures">
                  <Typography
                    style={{
                      color: `${dhwColorFinder(dhw)}`,
                      position: "relative",
                      bottom: 0,
                    }}
                    variant="h6"
                    color="textSecondary"
                  >
                    DHW
                  </Typography>
                </Tooltip>
              </Grid>
            </Grid>
          </Grid>
          <Grid style={{ margin: "1rem 0 1rem 0" }} container item xs={12}>
            <Grid item>
              <Link
                style={{ color: "inherit", textDecoration: "none" }}
                to={`/reefs/${reef.id}`}
              >
                <Button size="small" variant="outlined" color="primary">
                  EXPLORE
                </Button>
              </Link>
            </Grid>
          </Grid>
        </CardContent>
      </Card>
    </LeafletPopup>
  );
};

const styles = (theme: Theme) =>
  createStyles({
    popupHeader: {
      backgroundColor: theme.palette.primary.main,
    },

    popupHeaderContent: {
      width: "100%",
      overflowWrap: "break-word",
    },

    subheader: {
      color: "white",
    },

    popup: {
      width: 215,
    },
  });

interface PopupIncomingProps {
  reef: Reef;
  // Dictates whether the popup automatically opens when the reef is selected (reef on map is set)
  autoOpen?: boolean;
}

Popup.defaultProps = {
  autoOpen: true,
};

type PopupProps = PopupIncomingProps & WithStyles<typeof styles>;

export default withStyles(styles)(Popup);
