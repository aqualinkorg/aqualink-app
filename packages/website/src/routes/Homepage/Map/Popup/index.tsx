import React from "react";
import {
  withStyles,
  WithStyles,
  createStyles,
  Theme,
  Card,
  CardHeader,
  CardContent,
  Grid,
  Typography,
  Button,
} from "@material-ui/core";
import { Link } from "react-router-dom";

import { Reef } from "../../../../store/Reefs/types";
import { colors } from "../../../../layout/App/theme";
import { formatNumber } from "../../../../helpers/numberUtils";
import {
  colorFinder,
  degreeHeatingWeeksCalculator,
} from "../../../../helpers/degreeHeatingWeeks";

const Popup = ({ reef, classes }: PopupProps) => {
  const { degreeHeatingDays, maxBottomTemperature } =
    reef.latestDailyData || {};
  return (
    <Card>
      <CardHeader
        className={classes.popupHeader}
        title={reef.name}
        subheader={reef.region}
      />
      <CardContent>
        <Grid container item xs={12}>
          <Grid item xs={6}>
            <Grid container justify="flex-start" item xs={12}>
              <Typography variant="caption" color="textSecondary">
                {`TEMP AT ${reef.depth}M`}
              </Typography>
            </Grid>
            <Grid container justify="flex-start" item xs={12}>
              <Typography
                style={{ color: colors.lightBlue }}
                variant="h5"
                color="textSecondary"
              >
                {formatNumber(maxBottomTemperature, 1) || "- -"} &#8451;
              </Typography>
            </Grid>
          </Grid>
          <Grid item xs={6}>
            <Grid container justify="flex-end" item xs={12}>
              <Typography variant="caption" color="textSecondary">
                DEG. HEAT. DAYS
              </Typography>
            </Grid>
            <Grid container justify="flex-end" item xs={12}>
              <Typography
                style={{
                  color: `${colorFinder(
                    degreeHeatingWeeksCalculator(degreeHeatingDays)
                  )}`,
                }}
                variant="h5"
                color="textSecondary"
              >
                {formatNumber(degreeHeatingDays, 1)}
              </Typography>
            </Grid>
          </Grid>
        </Grid>
        <Grid
          style={{ margin: "1rem 0 1rem 0" }}
          container
          justify="flex-start"
          item
          xs={12}
        >
          <Grid item>
            <Link
              style={{ color: "inherit", textDecoration: "none" }}
              to={`/reefs/${reef.id}`}
            >
              <Button size="small" variant="outlined" color="primary">
                EXPLORE
              </Button>
            </Link>
          </Grid>
        </Grid>
      </CardContent>
    </Card>
  );
};

const styles = (theme: Theme) =>
  createStyles({
    popupHeader: {
      backgroundColor: theme.palette.primary.main,
    },
  });

interface PopupIncomingProps {
  reef: Reef;
}

type PopupProps = PopupIncomingProps & WithStyles<typeof styles>;

export default withStyles(styles)(Popup);
