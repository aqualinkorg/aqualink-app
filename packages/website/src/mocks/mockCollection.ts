import { CollectionDetails } from "../store/Collection/types";
import { mockReef } from "./mockReef";
import { mockUser } from "./mockUser";

export const mockCollection: CollectionDetails = {
  id: 1,
  isPublic: false,
  name: "Mock Collection",
  user: mockUser,
  reefs: [mockReef],
  reefIds: [1],
};
