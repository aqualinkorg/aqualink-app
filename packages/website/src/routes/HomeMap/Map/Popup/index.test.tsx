import React from "react";
import { render } from "@testing-library/react";

import { BrowserRouter as Router } from "react-router-dom";
import Popup from ".";
import { mockReef } from "../../../../mocks/mockReef";

jest.mock("react-leaflet", () => ({
  __esModule: true,
  Popup: (props: any) =>
    jest.requireActual("react").createElement("mock-LeafletPopup", props),
}));

test("renders as expected", () => {
  const { container } = render(
    <Router>
      <Popup reef={mockReef} />
    </Router>
  );
  expect(container).toMatchSnapshot();
});
