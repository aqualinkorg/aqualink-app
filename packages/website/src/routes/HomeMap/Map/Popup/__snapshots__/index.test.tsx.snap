// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popup should render with given state from Redux store 1`] = `
<div>
  <mock-leafletpopup
    autopan="false"
    classname="Popup-popup-4"
    closebutton="false"
  >
    <mock-card>
      <div
        class="MuiCardHeader-root Popup-popupHeader-1"
      >
        <div
          class="MuiCardHeader-content Popup-popupHeaderContent-2"
        >
          <mock-typography
            classname="MuiCardHeader-title"
            component="span"
            display="block"
            variant="h5"
          >
            <span>
              Hawaii
            </span>
          </mock-typography>
        </div>
      </div>
      <mock-cardcontent>
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-grid-xs-12"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
          >
            <div
              class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-grid-xs-12"
            >
              <mock-typography
                color="textSecondary"
                variant="caption"
              >
                SST
              </mock-typography>
            </div>
            <div
              class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-grid-xs-12"
            >
              <mock-typography
                color="textSecondary"
                style="color: rgb(22, 141, 189);"
                variant="h5"
              >
                15.9  °C
              </mock-typography>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
          >
            <div
              class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-grid-xs-12"
            >
              <mock-typography
                color="textSecondary"
                variant="caption"
              >
                HEAT STRESS
              </mock-typography>
            </div>
            <div
              class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-align-items-xs-flex-end MuiGrid-grid-xs-12"
            >
              <mock-typography
                color="textSecondary"
                style="color: rgb(8, 91, 163);"
                variant="h5"
              >
                0.0
                 
              </mock-typography>
              <mock-tooltip
                title="Degree Heating Weeks - a measure of the amount of time above the 20 year historical maximum temperatures"
              >
                <mock-typography
                  color="textSecondary"
                  style="color: rgb(8, 91, 163); position: relative; bottom: 0px;"
                  variant="h6"
                >
                  DHW
                </mock-typography>
              </mock-tooltip>
            </div>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-grid-xs-12"
          style="margin: 1rem 0px 1rem 0px;"
        >
          <div
            class="MuiGrid-root MuiGrid-item"
          >
            <a
              href="/reefs/1"
              style="text-decoration: none;"
            >
              <mock-button
                color="primary"
                size="small"
                variant="outlined"
              >
                EXPLORE
              </mock-button>
            </a>
          </div>
        </div>
      </mock-cardcontent>
    </mock-card>
  </mock-leafletpopup>
</div>
`;
